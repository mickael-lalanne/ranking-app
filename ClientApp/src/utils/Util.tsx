import React from "react";

// All items that are not yet saved in the database starts with the following identifier
const TMP_ID_IDENTIFIER = '99999';

/**
 * Generate a temporary for items that are not yet saved in database
 * @returns {string} a random temporary id
 */
export const generateRandomId = (): string => {
    const randomNumber: number = Math.floor(Math.random() * 1000000000);
    return TMP_ID_IDENTIFIER + randomNumber;
};

/**
 * Check if an id is temporary
 * @param {string} id the id to check
 * @returns {boolean} true if the id is temporary
 */
export const isTemporaryId = (id: string): boolean => {
    return id.startsWith(TMP_ID_IDENTIFIER);
}

export interface ResizedImage {
    source: string; // the data URL of the image
    name: string;
}

/**
 * Resize an image while keeping its ratio
 * This function is called in the "onChange" event of input file
 * So we use the FileReader api to get the base image dimensions
 * @param {File} imageFile the image file to resize
 * @returns {Promise<ResizedImage>} an object containing the resized image as data URL and its name
 */
export const resizeImage = (imageFile: File): Promise<ResizedImage> => {
    const MAX_SIZE: number = 150;
    const reader = new FileReader();

    return new Promise ((resolve) => {
        reader.onload = () => {
            const tmpImg: HTMLImageElement = new Image();
    
            tmpImg.onload = () => {
                const resizeRatio = Math.min(MAX_SIZE / tmpImg.width, MAX_SIZE / tmpImg.height);

                // Initialize the canvas
                const canvas = document.createElement("canvas");
                const ctx = canvas.getContext('2d');
    
                // Set width and height
                canvas.width = tmpImg.width * resizeRatio;
                canvas.height = tmpImg.height * resizeRatio;
    
                // Draw image and export to a data-uri
                ctx!.drawImage(tmpImg, 0, 0, canvas.width, canvas.height);
    
                const dataURI = canvas.toDataURL();
                resolve({ source: dataURI, name: imageFile.name });
            };

            tmpImg.src = URL.createObjectURL(imageFile);
        }
        reader.readAsDataURL(imageFile);
    });

};

/**
 * @param array an array containing a createdAt field
 * @returns {Partial<{ createdAt: string }>[]} the same array but sorted by creation date
 */
export const sortByCreationDate = (array: Partial<{ createdAt: string }>[]) => {
    // Use slice() to avoid reference errors
    return array.slice().sort((a, b) => 
        new Date(a.createdAt!).getTime() - new Date(b.createdAt!).getTime()
    );
};

/**
 * Called in the template and tierlist editors
 * Returns a value for the MUI Tooltip component
 * @param {string?} content the text that should be displayed in the tooltip 
 * @returns {'' | React.JSX.Element | undefined} a span including the text if defined, otherwise undefined
 */
export const getTooltipTitleForSaveButtons = (
    content?: string
): '' | React.JSX.Element | undefined => {
    return content && <span style={{ fontSize: '15px'}}>{content}</span>;
}
