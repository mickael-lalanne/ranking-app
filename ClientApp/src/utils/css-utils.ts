export const ELEMENT_SIZE: string = '75px';
