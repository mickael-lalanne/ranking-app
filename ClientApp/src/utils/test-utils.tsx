import React, { PropsWithChildren } from 'react';
import { render } from '@testing-library/react';
import type { RenderOptions } from '@testing-library/react';
import { Provider } from 'react-redux';
import { setupStore } from '../app/store';
import type { AppStore, RootState } from '../app/store';
import { ThemeOptions, ThemeProvider, createTheme } from '@mui/material/styles';
import '@testing-library/jest-dom';
import { DndProvider } from 'react-dnd';
import { TouchBackend } from 'react-dnd-touch-backend';

// This type interface extends the default options for render from RTL, as well
// as allows the user to specify other things such as initialState, store.
export interface ExtendedRenderOptions extends Omit<RenderOptions, 'queries'> {
    preloadedState?: Partial<RootState>;
    store?: AppStore;
}

// Mock MUI theme
const RANKING_APP_THEME_MOCK: ThemeOptions = {
    defaultRankingTheme: {
        primary: '',
        primaryLight: '',
        light: '',
        dark: '',
        info: '',
    },
};
const MockTheme = ({ children }: any) => {
    const theme = createTheme(RANKING_APP_THEME_MOCK);
    return <ThemeProvider theme={theme}>{children}</ThemeProvider>;
};

export function renderWithProviders(
    ui: React.ReactElement,
    {
        preloadedState = {},
        // Automatically create a store instance if no store was passed in
        store = setupStore(preloadedState),
        ...renderOptions
    }: ExtendedRenderOptions = {}
) {
    function Wrapper({ children }: PropsWithChildren<{}>): JSX.Element {
        return (
            <MockTheme>
                <Provider store={store}>
                    <DndProvider
                        backend={TouchBackend}
                        options={{ enableMouseEvents: true }}
                    >
                        {children}
                    </DndProvider>
                </Provider>
            </MockTheme>
        );
    }

    // Return an object with the store and all of RTL's query functions
    return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) };
}
