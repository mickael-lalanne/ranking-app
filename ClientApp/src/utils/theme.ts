import { Theme, createTheme, alpha, ThemeOptions } from "@mui/material";

interface RankingTheme {
    primary: string;
    primaryLight: string;
    light: string;
    dark: string;
    info: string;
};
const PRIMARY_COLOR = '#BB2525';

declare module '@mui/material/styles' {
    interface Theme {
        defaultRankingTheme: {
            primary: string;
            primaryLight: string;
            light: string;
            dark: string;
            info: string;
        };
    }
    // allow configuration using `createTheme`
    interface ThemeOptions {
        defaultRankingTheme?: {
            primary: string;
            primaryLight: string;
            light: string;
            dark: string;
            info: string;
        };
    }
}

export const THEMES_COLORS: ThemeOptions['defaultRankingTheme'] = {
    primary: PRIMARY_COLOR,
    primaryLight: '#FF6969',
    light: '#FFF5E0',
    dark: '#141E46',
    info: '#9ED9EB'
}

export const RANKING_APP_THEME: Theme = createTheme({
    defaultRankingTheme: THEMES_COLORS,
    palette: {
        primary: {
            main: alpha(PRIMARY_COLOR, 0.7),
            light: alpha(PRIMARY_COLOR, 0.5),
            dark: alpha(PRIMARY_COLOR, 0.9),
        }
    }
});
