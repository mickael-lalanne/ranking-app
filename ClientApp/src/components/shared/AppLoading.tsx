import React from 'react';
import { css, keyframes } from '@emotion/css';
import { THEMES_COLORS } from '../../utils/theme';

const AppLoading = ({ text }: { text: string; }) => {
    return (<div className={container_style} data-cy="app-loading">
        <div className={scene_style}>
            <div className={cube_wrapper_style}>
                <div className={cube_style}>
                    <div className={cube_faces_style}>
                        <div className={cube_face_style + ' ' + shadow_style}></div>
                        <div className={cube_face_style + ' ' + bottom_style}></div>
                        <div className={cube_face_style + ' ' + top_style}></div>
                        <div className={cube_face_style + ' ' + left_style}></div>
                        <div className={cube_face_style + ' ' + right_style}></div>
                        <div className={cube_face_style + ' ' + back_style}></div>
                        <div className={cube_face_style + ' ' + front_style}></div>
                    </div>
                </div>
            </div>
        </div>
        <div className={text_container_style}>
            <div className={text_style}>{text}</div>
            <div className={dot_pulse_style}></div>
        </div>
    </div>);
};

export default AppLoading;

/**
 * CSS STYLES
 */
const scale = keyframes`
    0% {
        transform: scale(0);
    }
    100% {
        transform: scale(1);
    }
`;

const container_style = css({
    width: 'fit-content',
    animation: `${scale} .5s`,
    height: 'fit-content'
});

const text_container_style = css({
    display: 'flex',
    alignItems: 'center',
    width: 'fit-content'
});

const text_style = css({
    marginRight: '30px',
    fontFamily: '"Raleway", sans-serif',
    color: THEMES_COLORS?.primary,
    fontSize: '18px',
    fontWeight: 600,
    fontStyle: 'italic'
});

/****************************************
 *           CUBE ANIMATION
 ****************************************/
const bouncing = keyframes`
    0% {
        transform: translateY(-40px);
        animation-timing-function: cubic-bezier(0.76, 0.05, 0.86, 0.06);
    }
    45% {
        transform: translateY(40px);
        animation-timing-function: cubic-bezier(0.23, 1, 0.32, 1);
    }
    100% {
        transform: translateY(-40px);
        animation-timing-function: cubic-bezier(0.76, 0.05, 0.86, 0.06);
    }
`;

const bouncingShadow = keyframes`
    0% {
        transform: translateZ(-80px) scale(1.3);
        animation-timing-function: cubic-bezier(0.76, 0.05, 0.86, 0.06);
        opacity: 0.05;
    }
    45% {
        transform: translateZ(0);
        animation-timing-function: cubic-bezier(0.23, 1, 0.32, 1);
        opacity: 0.3;
    }
    100% {
        transform: translateZ(-80px) scale(1.3);
        animation-timing-function: cubic-bezier(0.76, 0.05, 0.86, 0.06);
        opacity: 0.05;
    }
`;

const rotation = keyframes`
    0% {
        transform: rotateX(45deg) rotateY(0) rotateZ(45deg);
        animation-timing-function: cubic-bezier(0.17, 0.84, 0.44, 1);
    }
    50% {
        transform: rotateX(45deg) rotateY(0) rotateZ(225deg);
        animation-timing-function: cubic-bezier(0.76, 0.05, 0.86, 0.06);
    }
    100% {
        transform: rotateX(45deg) rotateY(0) rotateZ(405deg);
        animation-timing-function: cubic-bezier(0.17, 0.84, 0.44, 1);
    }
`;

const scene_style = css({
    position: 'relative',
    zIndex: 2,
    height: '220px',
    width: '220px',
    display: 'grid',
    placeItems: 'center'
});

const cube_wrapper_style = css({
    transformStyle: 'preserve-3d',
    animation: `${bouncing} 2s infinite`
});

const cube_style = css({
    transformStyle: 'preserve-3d',
    transform: 'rotateX(45deg) rotateZ(45deg)',
    animation: `${rotation} 2s infinite`
});

const cube_faces_style = css({
    transformStyle: 'preserve-3d',
    height: '80px',
    width: '80px',
    position: 'relative',
    transformOrigin: '0 0',
    transform: 'translateX(0) translateY(0) translateZ(-40px)'
});

const cube_face_style = css({
    position: 'absolute',
    inset: 0,
    background: THEMES_COLORS?.dark,
    border: `solid 1px ${THEMES_COLORS?.primary}`
});

const shadow_style = css({
    transform: 'translateZ(-80px)',
    animation: `${bouncingShadow} 2s infinite`
});

const bottom_style = css({
});

const top_style = css({
    transform: 'translateZ(80px)'
});

const left_style = css({
    transformOrigin: '50% 0',
    transform: 'rotateX(-90deg) translateY(-80px) translateZ(80px)'
});

const right_style = css({
    transformOrigin: '50% 0',
    transform: 'rotateX(-90deg) translateY(-80px)'
});

const back_style = css({
    transformOrigin: '0 50%',
    transform: 'rotateY(-90deg) translateZ(-80px)'
});

const front_style = css({
    transformOrigin: '0 50%',
    transform: 'rotateY(-90deg)'
});

/****************************************
 *           THREE DOTS
 ****************************************/

const dotPulse = keyframes`
    0% {
        box-shadow: 9999px 0 0 -5px;
    }
    30% {
        box-shadow: 9999px 0 0 2px;
    }
    60%, 100% {
        box-shadow: 9999px 0 0 -5px;
    }
`;

const dotPulseBefore = keyframes`
    0% {
        box-shadow: 9984px 0 0 -5px;
    }
    30% {
        box-shadow: 9984px 0 0 2px;
    }
    60%, 100% {
        box-shadow: 9984px 0 0 -5px;
    }
`;

const dotPulseAfter = keyframes`
    0% {
        box-shadow: 10014px 0 0 -5px;
    }
    30% {
        box-shadow: 10014px 0 0 2px;
    }
    60%, 100% {
        box-shadow: 10014px 0 0 -5px;
    }
`;

const dot_pulse_style = css({
    position: 'relative',
    left: '-9999px',
    width: '10px',
    height: '10px',
    borderRadius: '5px',
    backgroundColor: THEMES_COLORS?.primary,
    color: THEMES_COLORS?.primary,
    boxShadow: '9999px 0 0 -5px',
    animation: `${dotPulse} 1.5s infinite linear`,
    animationDelay: '0.25s',
    ':before, :after': {
        content: '""',
        display: 'inline-block',
        position: 'absolute',
        top: 0,
        width: '10px',
        height: '10px',
        borderRadius: '5px',
        backgroundColor: THEMES_COLORS?.primary,
        color: THEMES_COLORS?.primary
    },
    ':before': {
        boxShadow: '9984px 0 0 -5px',
        animation: `${dotPulseBefore} 1.5s infinite linear`,
        animationDelay: '0s'
    },
    ':after': {
        boxShadow: '10014px 0 0 -5px',
        animation: `${dotPulseAfter} 1.5s infinite linear`,
        animationDelay: '0.5s'
    }
});
