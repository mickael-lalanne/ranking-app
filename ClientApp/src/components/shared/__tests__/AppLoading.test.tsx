import * as React from 'react';
import { renderWithProviders } from '../../../utils/test-utils';
import { screen } from '@testing-library/react';
import AppLoading from '../AppLoading';

const loadingText: string = 'My loading indicator';

const AppLoadingTest = () => {
    return <AppLoading text={loadingText} />;
};

describe('Screenshot', () => {
    it('takes default screenshot', () => {
        const component = renderWithProviders(AppLoadingTest());

        expect(component.baseElement).toMatchSnapshot();
    });
});

describe('Common', () => {
    it('checks the text is displayed', () => {
        renderWithProviders(AppLoadingTest());

        expect(screen.getByText(loadingText)).toBeDefined();
    });
});
