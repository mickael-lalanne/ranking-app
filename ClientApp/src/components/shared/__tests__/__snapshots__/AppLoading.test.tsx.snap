// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Screenshot takes default screenshot 1`] = `
<body>
  <div>
    <div
      class="css-1yhl6zy"
      data-cy="app-loading"
    >
      <div
        class="css-12lhfml"
      >
        <div
          class="css-fvsue4"
        >
          <div
            class="css-qls6c6"
          >
            <div
              class="css-hthcbb"
            >
              <div
                class="css-1b6vs5j css-8q6na4"
              />
              <div
                class="css-1b6vs5j css-0"
              />
              <div
                class="css-1b6vs5j css-1t5wwna"
              />
              <div
                class="css-1b6vs5j css-1gu2kcg"
              />
              <div
                class="css-1b6vs5j css-17zyqrd"
              />
              <div
                class="css-1b6vs5j css-156lazp"
              />
              <div
                class="css-1b6vs5j css-17i4z6a"
              />
            </div>
          </div>
        </div>
      </div>
      <div
        class="css-xyomv4"
      >
        <div
          class="css-ircun"
        >
          My loading indicator
        </div>
        <div
          class="css-1lfmw4p"
        />
      </div>
    </div>
  </div>
</body>
`;
